Updates to the data layout chapter (#96)
rust-lang-nursery/nomicon: src/other-reprs.md

# repr(transparent)

> This can be put only on structs with just one non-zero-sized field. The effect

Reviewer: ```suggestion
This can only be used on structs with a single non-zero-sized field. The effect
```

> is, the representation of the whole struct is guaranteed to be the same as that

Reviewer: ```suggestion
is that the layout and ABI of the whole struct is guaranteed to be the same as that
```

one field.

The goal is to make it possible to transmute between the single field and the
struct. An example of that is the [`UnsafeCell`], which can be transmuted into

> the wrapped thing.

Reviewer: ```suggestion
the type it wraps.
```

Also, passing the struct through FFI where the inner field type is expected on

> the other side is allowed.

Reviewer: ```suggestion
the other side is allowed. In particular, this is necessary for `struct Meters(f32)`
to have the same ABI as `f32`.
```

However if you take a reference to a packed field, it's unlikely that the
compiler will be able to emit code to avoid an unaligned load.

**[As of Rust 1.30.0 this still can cause undefined behavior.][ub loads]**

`repr(packed)` is not to be used lightly. Unless you have extreme requirements,
this should not be used.

This repr is a modifier on `repr(C)` and `repr(rust)`.

# repr(align(n))

> `repr(align(n))` (where `n` is a power of two) forces the type to have alignment

Reviewer: ```suggestion
`repr(align(n))` (where `n` is a power of two) forces the type to have an alignment of
```
